### ISSUE TITLE
Broken union expansion in overload matching

### ISSUE THREAD

**Issue by viccie30** [microsoft/pyright#4892]
**Describe the bug**
The code from #1298 again gives an error:
```
reproducer.py
  reproducer.py:5:15 - error: Argument of type "str | bytes | PathLike[str] | PathLike[bytes]" cannot be assigned to parameter "path" of type "PathLike[AnyStr@fspath]" in function "fspath"
    Type "str | bytes | PathLike[str] | PathLike[bytes]" cannot be assigned to type "PathLike[AnyStr@fspath]"
      "bytes" is incompatible with protocol "PathLike[AnyStr@fspath]"
        "__fspath__" is not present (reportGeneralTypeIssues)
```

**To Reproduce**
```shell
cat << _EOF > reproducer.py
import os
from typing import Union

def foo(path: Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]) -> None:
    os.fspath(path)
_EOF
pyright reproducer.py
```

**Expected behavior**
I expected the type check to sucee

**Screenshots or Code**
```python
import os
from typing import Union

def foo(path: Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]) -> None:
    os.fspath(path)
```

**VS Code extension or command-line**
pyright CLI 1.1.301

**Additional context**
No context, just thanks for a great piece of software!

**Comment by erictraut** [comment]
I've confirmed that this change in behavior started with [this commit](https://github.com/microsoft/pyright/commit/ba90156fc948fc31aa95c2ee9eb2c0862bbb6d47), which was introduced in version 1.1.262 of pyright (yes, 39 versions and nearly a year ago!). I'll investigate further.

**Comment by erictraut** [comment]
The overload behavior in pyright has gone through extensive changes since 1.1.262. This includes not only some behavior changes but also some important optimizations.

I think the current behavior is arguably correct.

Let's take a simpler example that doesn't involve any overloads.

```python
def fspath(path: PathLike[AnyStr]) -> AnyStr:
    ...

def foo(path: PathLike[str] | PathLike[bytes]) -> None:
    fspath(path) # Type error
```

In this case, both pyright and mypy correctly report an error when attempting to call `fspath`.

Now, let's examine the overloaded case.

```python
@overload
def fspath(path: str) -> str: ...

@overload
def fspath(path: bytes) -> bytes: ...

@overload
def fspath(path: PathLike[AnyStr]) -> AnyStr: ...

def foo(path: PathLike[str] | PathLike[bytes]) -> None:
    fspath(path)
```

Interestingly, mypy no longer reports an error in this case. Pyright does, which I think is consistent with the non-overload case.

The way I suggest handling this in your code — a technique that will work with both mypy and pyright — is to use `PathLike[AnyStr]` in the type definition for the `path` parameter.

```python
def foo(path: str | bytes | PathLike[AnyStr]) -> None:
    fspath(path)
```
